### PR TITLE
OIDC provider

### PR DESCRIPTION
**Pull request by afabiani**
This Pull Request introduces the OpenID Connect Provider Filter for GeoStore

### CODE REVIEW

**Comment by tdipisa** [comment]
@afabiani can you please fix conflicts so that we can proceed finally with the review?

**Comment by afabiani** [comment]
> @afabiani can you please fix conflicts so that we can proceed finally with the review?

@tdipisa  done

File: src/modules/rest/impl/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTServiceImpl.java [hunk 1]
```diff
@@ -56,31 +46,48 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
 /**
  * Class RESTServiceImpl.
- * 
- * This is the super class for each RESTServices implementation 
- * 
+ * <p>
```
Comment (offtherailz) [inline src/modules/rest/impl/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTServiceImpl.java:58]:
```suggestion
 * 
```

**Review by offtherailz — CHANGES_REQUESTED** [review]
At the end of the review I found that only 9 files have significant changes.
Not any particular breaking change, a generic OIDC connect has been added to the current configuration.
So also this PR is safe to merge and backport. :+1: nice 

It doesn't introduce the possibility of configure multiple ones (now we can have one google, one keycloak, one generic oidc. I think that anyway this is enough).

Is not clear to me how to configure this to test. Looking at various superclasses I found this list of properties:
- IdPConfiguration
    -  protected String `beanName`;
    - protected boolean `enabled` = false;
    - protected boolean `autoCreateUser`;
    - protected String `internalRedirectUri`;
    -  protected String `redirectUri`;
    - protected Role `authenticatedDefaultRole`;
- Oauth2Configuration
     - protected String `clientId`;
     - protected String `clientSecret`;
     - protected String `accessTokenUri`;
     - protected String `authorizationUri`;
     - protected String `checkTokenEndpointUrl`;
     - protected String `logoutUri`;
     - protected String `scopes`;
     - protected String `idTokenUri`;
     - protected String `discoveryUrl`;
     - protected String `revokeEndpoint`;
     - protected boolean `enableRedirectEntryPoint` = false;
     - protected String `principalKey`;
     - protected String `rolesClaim`;
     - protected String `groupsClaim`;
- OpenIdConnectConfiguration
   - `jwkURI`
   - `responseMode`
   - `postLogoutRedirectUri`
   - `sendClientSecret`
   - `allowBearerToken`
   - `usePKCE`

Here my requests: 

- [ ] `sendClientSecret`, `usePKCE` are not clear, please document them in the JSDoc. 
- [ ] `responseMode` and `allowBearerToken` don't look to be used. Please remove them
- [ ] `jwkURI` too doesn't seems to be used, only used to set the `store` property of `GeoStoreOauthRestTemplate` but never used again. Also this store is private and the PR introduces it without adding any access to the private property or propertyAccessor. If it is not needed, please remove it. 
- [ ] Please provide a sample configuration and instructions to test it.

**Comment by afabiani** [comment]
@offtherailz please see below:

- `responseMode` and `allowBearerToken` ok to be removed
- `sendClientSecret` and `usePKCE` ok to be documented
- `jwkURI` should contain the url for the JWT SHA keys validation. This is used by the token store automatically. The validation is optional but strongly recommended. The `store` will be instantiated accordingly and used to validate the JWT token.
- All the parameters you listed above must be set in order to correctly configure the `oidc` connect. Most of them are tipically available through the `.well_knwon` endpoints

**Comment by tdipisa** [comment]
@afabiani is this ready for a new review or are you still working on it?

**Comment by afabiani** [comment]
Nope, no other additions on my side

**Comment by offtherailz** [comment]
trying to set it up in mapstore I received the following error:

Alessio, do you know how to fix it? 
```
Offending resource: URL [jar:file:/home/offtherailz/work/projects/MapStore2/product/target/apache-tomcat-8.5.69/webapps/mapstore/WEB-INF/lib/geostore-rest-impl-2.2-SNAPSHOT.jar!/applicationContext.xml]; nested exception is org.springframework.beans.factory.parsing.BeanDefinitionParsingException: Configuration problem: Filter beans '<oidcOpenIdFilter>' and '<googleOpenIdFilter>' have the same 'order' value. When using custom filters, please make sure the positions do not conflict with default filters. Alternatively you can disable the default filters by removing the corresponding child elements from <http> and avoiding the use of <http auto-config='true'>.
Offending resource: URL [file:/home/offtherailz/work/projects/MapStore2/product/target/apache-tomcat-8.5.69/webapps/mapstore/WEB-INF/classes/geostore-spring-security.xml]
	at org.springframework.beans.factory.parsing.FailFastProblemReporter.error(FailFastProblemReporter.java:72) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.beans.factory.parsing.ReaderContext.error(ReaderContext.java:119) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.beans.factory.parsing.ReaderContext.error(ReaderContext.java:104) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.beans.factory.xml.DefaultBeanDefinitionDocumentReader.importBeanDefinitionResource(DefaultBeanDefinitionDocumentReader.java:240) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.beans.factory.xml.DefaultBeanDefinitionDocumentReader.parseDefaultElement(DefaultBeanDefinitionDocumentReader.java:191) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.beans.factory.xml.DefaultBeanDefinitionDocumentReader.parseBeanDefinitions(DefaultBeanDefinitionDocumentReader.java:176) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.beans.factory.xml.DefaultBeanDefinitionDocumentReader.doRegisterBeanDefinitions(DefaultBeanDefinitionDocumentReader.java:149) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.beans.factory.xml.DefaultBeanDefinitionDocumentReader.registerBeanDefinitions(DefaultBeanDefinitionDocumentReader.java:96) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.registerBeanDefinitions(XmlBeanDefinitionReader.java:511) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.doLoadBeanDefinitions(XmlBeanDefinitionReader.java:391) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.loadBeanDefinitions(XmlBeanDefinitionReader.java:338) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.loadBeanDefinitions(XmlBeanDefinitionReader.java:310) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.beans.factory.support.AbstractBeanDefinitionReader.loadBeanDefinitions(AbstractBeanDefinitionReader.java:196) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.beans.factory.support.AbstractBeanDefinitionReader.loadBeanDefinitions(AbstractBeanDefinitionReader.java:232) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.beans.factory.support.AbstractBeanDefinitionReader.loadBeanDefinitions(AbstractBeanDefinitionReader.java:203) ~[spring-beans-5.3.18.jar:5.3.18]
	at org.springframework.web.context.support.XmlWebApplicationContext.loadBeanDefinitions(XmlWebApplicationContext.java:125) ~[spring-web-5.3.18.jar:5.3.18]
	at org.springframework.web.context.support.XmlWebApplicationContext.loadBeanDefinitions(XmlWebApplicationContext.java:94) ~[spring-web-5.3.18.jar:5.3.18]
	at org.springframework.context.support.AbstractRefreshableApplicationContext.refreshBeanFactory(AbstractRefreshableApplicationContext.java:130) ~[spring-context-5.3.18.jar:5.3.18]
	at org.springframework.context.support.AbstractApplicationContext.obtainFreshBeanFactory(AbstractApplicationContext.java:671) ~[spring-context-5.3.18.jar:5.3.18]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:553) ~[spring-context-5.3.18.jar:5.3.18]
	at org.springframework.web.context.ContextLoader.configureAndRefreshWebApplicationContext(ContextLoader.java:401) ~[spring-web-5.3.18.jar:5.3.18]
	at org.springframework.web.context.ContextLoader.initWebApplicationContext(ContextLoader.java:292) ~[spring-web-5.3.18.jar:5.3.18]
	at org.springframework.web.context.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:103) ~[spring-web-5.3.18.jar:5.3.18]
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4763) ~[catalina.jar:8.5.69]
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5232) ~[catalina.jar:8.5.69]
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183) ~[catalina.jar:8.5.69]
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:755) ~[catalina.jar:8.5.69]
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:729) ~[catalina.jar:8.5.69]
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:695) ~[catalina.jar:8.5.69]
	at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:1016) ~[catalina.jar:8.5.69]
	at org.apache.catalina.startup.HostConfig$DeployWar.run(HostConfig.java:1903) ~[catalina.jar:8.5.69]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:829) ~[?:?]
```

**Comment by afabiani** [comment]
@offtherailz try to update the `geostore-spring-security.xml` like this

```xml
<security:http auto-config="true" create-session="never" >
		<security:http-basic entry-point-ref="restAuthenticationEntryPoint"/>
		<security:csrf disabled="true"/>
		<security:custom-filter ref="authenticationTokenProcessingFilter" before="FORM_LOGIN_FILTER"/>
		<security:custom-filter ref="sessionTokenProcessingFilter" after="FORM_LOGIN_FILTER"/>
		<security:custom-filter ref="keycloakFilter" before="BASIC_AUTH_FILTER"/>
		<security:custom-filter ref="googleOpenIdFilter" after="BASIC_AUTH_FILTER"/>
		<security:custom-filter ref="oidcOpenIdFilter" before="OPENID_FILTER"/>
		<security:anonymous />
	</security:http>
```

File: src/web/app/src/main/resources/geostore-spring-security.xml [hunk 2]
```diff
@@ -27,6 +27,7 @@
 		<security:custom-filter ref="sessionTokenProcessingFilter" after="FORM_LOGIN_FILTER"/>
 		<security:custom-filter ref="keycloakFilter" before="BASIC_AUTH_FILTER"/>
 		<security:custom-filter ref="googleOpenIdFilter" after="BASIC_AUTH_FILTER"/>
+		<security:custom-filter ref="oidcOpenIdFilter" after="BASIC_AUTH_FILTER"/>
```
Comment (offtherailz) [inline src/web/app/src/main/resources/geostore-spring-security.xml:30]:
```suggestion
		<security:custom-filter ref="oidcOpenIdFilter" before="OPENID_FILTER"/>
```

**Comment by offtherailz** [comment]
@afabiani notes for testing


## Testing

```
docker pull quay.io/keycloak/keycloak:latest

docker run --name keycloak_server -p 8180:8180 \ -e KEYCLOAK_ADMIN=admin \ -e KEYCLOAK_ADMIN_PASSWORD=password \ quay.io/keycloak/keycloak:latest \ start-dev \ --http-port=8180
```

keycloak new realm


Save 

- disable Direct access grants
- enable Client authentication
Save
- copy  Credentials --> Client secret
	- Testare l'url http://localhost:8180/realms/mapstore/.well-known/openid-configuration 

create a file: 
`datadir/mapstore-ovr.properties`
```properties
# Client Secret xBRWNDXiAC4fPp8l89hGCBY6QGDPDDqa

# enables the keycloak OpenID Connect filter

oidcOAuth2Config.enabled=true

# note: this is the client id you have created in Keycloak

oidcOAuth2Config.clientId=mapstore

oidcOAuth2Config.clientSecret=xBRWNDXiAC4fPp8l89hGCBY6QGDPDDqa

oidcOAuth2Config.discoveryUrl=http://localhost:8180/realms/mapstore/.well-known/openid-configuration

oidcOAuth2Config.sendClientSecret=true

  

oidcOAuth2Config.redirectUri=http://localhost:8081/rest/geostore/openid/oidc/callback

# Internal redirect URI (you can set it to relative path like this `../../..` to make this config work across domain)

oidcOAuth2Config.internalRedirectUri=https://localhost:8081/
```

# MapStore debug


edit `product/pom.xml` to configure data dir
```diff
# product/pom.xml
<containerId>tomcat8x</containerId>

+ <systemProperties>
+ <datadir.location>${env.DATADIR}</datadir.location>
+ </systemProperties>
```



open 2 consoles
1) mapstore front-end
```bash
# MapStore front-end
export MAPSTORE_BACKEND_PORT=8082
npm run fe:start
```

2) backend
```
export MAPSTORE_BACKEND_PORT=8082
## Change with path to your data dir.
export DATADIR=/home/offtherailz/work/projects/MapStore2/datadir 
# debug (opt)
# export MAVEN_OPTS="-Xdebug -Xnoagent -Djava.compiler=NONE -Xrunjdwp:transport=dt_socket,address=4000,server=y,suspend=n"
```

**Comment by afabiani** [comment]
@offtherailz see [here](https://github.com/geosolutions-it/support/issues/2408#issuecomment-2028015775)

**Review by offtherailz — APPROVED** [review]
This integration works well and satisfies the minimum requirement to integrate login. Tested with keycloak.

**Comment by tdipisa** [comment]
@offtherailz @afabiani, once tested and accepted, this need to be backported to 2.1.x branch for [MS 2024.01.01](https://github.com/geosolutions-it/MapStore2/wiki/MapStore-Release-Calendars#release-calendar-2024)
@offtherailz I suppose there is something to be included in the MS dev guide for this. Isn't it? Please confirm.

**Comment by offtherailz** [comment]
@tdipisa this was assigned to 2024.02.00 as far as I see
I also created https://github.com/geosolutions-it/MapStore2/issues/10151 and https://github.com/geosolutions-it/MapStore2/pull/10152 in the same milestone. 
Please let us know if there is a change in plans.

**Comment by tdipisa** [comment]
> @tdipisa this was assigned to 2024.02.00 as far as I see I also created [geosolutions-it/MapStore2#10151](https://github.com/geosolutions-it/MapStore2/issues/10151) and [geosolutions-it/MapStore2#10152](https://github.com/geosolutions-it/MapStore2/pull/10152) in the same milestone. Please let us know if there is a change in plans.

milestone changed according to https://github.com/geosolutions-it/geostore/pull/330#issuecomment-2034919898

**Pull request merged**